cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,804 @@
+                        if (response.thereIsNotEnoughDataForProposal()) {
+                            // We do not include a session id with this status as we do not want to retrieve the state of
+                            // this failed tasks (COMPLETED_WITH_ERROR)
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .addNewCondition().withNewType(State.PendingProposal.toString()).endCondition().build();
+                        } else if (response.proposalIsStillCalculating()) {
+                            return new KafkaClusterRebalanceStatusBuilder()
+                                    .withNewSessionId(response.getUserTaskId())
+                                    .addNewCondition().withNewType(State.Rebalancing.toString()).endCondition().build();
+                        }
+                    }
+
+                    if (response.getJson().containsKey(CC_REST_API_SUMMARY)) {
+                        // If there is enough data and the proposal is complete (the response has the "summary" key) then we move
+                        // to ProposalReady for a dry run or to the Rebalancing state for a full run
+                        State ready = dryrun ? State.ProposalReady : State.Rebalancing;
+                        return new KafkaClusterRebalanceStatusBuilder()
+                                .withOptimizationResult(response.getJson().getJsonObject(CC_REST_API_SUMMARY).getMap())
+                                .addNewCondition().withNewType(ready.toString()).endCondition().build();
+                    } else {
+                        throw new RuntimeException("Rebalance returned unknown response: " + response.toString());
+                    }
+                }).otherwise(t -> {
+                    throw new RuntimeException(t.getMessage());
+                });
+    }
+
+    private RebalanceAnnotation getRebalanceAnnotation(KafkaClusterRebalance clusterRebalance) {
+        String rebalanceAnnotationValue = rebalanceAnnotation(clusterRebalance);
+        RebalanceAnnotation rebalanceAnnotation;
+        try {
+            rebalanceAnnotation = rebalanceAnnotationValue == null ?
+                    RebalanceAnnotation.none : RebalanceAnnotation.valueOf(rebalanceAnnotationValue);
+        } catch (IllegalArgumentException e) {
+            rebalanceAnnotation = RebalanceAnnotation.unknown;
+            log.warn("Wrong annotation value {}={} on {}/{}",
+                    ANNO_STRIMZI_IO_REBALANCE, rebalanceAnnotationValue,
+                    clusterRebalance.getMetadata().getNamespace(), clusterRebalance.getMetadata().getName());
+        }
+        return rebalanceAnnotation;
+    }
+
+    private String rebalanceAnnotation(KafkaClusterRebalance clusterRebalance) {
+        return clusterRebalance.getMetadata().getAnnotations() == null ?
+                null : clusterRebalance.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_REBALANCE);
+    }
```

Comment:
Do we really need this one?

Reply:
This takes the raw string value for the annotation which could even be a not valid value (i.e. `strimzi.io/rebalance=foo`). The other method maps to the corresponding annotation to `unknown` in this case (or `none` if no annotation is present). I renamed and added a few Javadoc on all of them.